tests/MongoDB.Driver.Tests/Specifications/transactions/TransactionTestRunner.cs

```diff
@@ -295,24 +356,50 @@ private ClientSessionOptions ParseSessionOptions(BsonDocument test, string sessi
             return options;
         }
 
-        private FailPoint ConfigureFailPoint(BsonDocument test)
+        private FailPoint ConfigureFailPoint(IServer server, BsonValue failPoint)
+        {
+            var session = NoCoreSession.NewHandle();
+            var command = failPoint.AsBsonDocument;
+            return FailPoint.Configure(server, session, command);
+
+        }
+
+        private DisposableBundle<FailPoint> ConfigureFailPointOnPrimaryOrShardRoutersIfNeeded(IMongoClient client, BsonDocument test)
         {
-            BsonValue failPoint;
-            if (test.TryGetValue("failPoint", out failPoint))
+            if (!test.TryGetValue("failPoint", out var failPoint))
             {
-                var cluster = DriverTestConfiguration.Client.Cluster;
-                var server = cluster.SelectServer(WritableServerSelector.Instance, CancellationToken.None);
-                var session = NoCoreSession.NewHandle();
-                var command = failPoint.AsBsonDocument;
-                return FailPoint.Configure(cluster, session, command);
+                return null;
             }
 
-            return null;
+            var cluster = client.Cluster;
+            var timeOut = TimeSpan.FromSeconds(60);
+            SpinWait.SpinUntil(() => cluster.Description.Type != ClusterType.Unknown, timeOut).Should().BeTrue();
+
+            IEnumerable<IServer> failPointServers;
+            switch (cluster.Description.Type)
+            {
+                case ClusterType.ReplicaSet:
+                    var primary = cluster.SelectServer(WritableServerSelector.Instance, CancellationToken.None);
+                    failPointServers = new[] {primary};
+                    break;
+                
+                case ClusterType.Sharded:
+                    failPointServers = cluster.Description.Servers
+                        .Select(server => server.EndPoint)
+                        .Select(endPoint => cluster.SelectServer(new EndPointServerSelector(endPoint), CancellationToken.None));
+                    break;
+                default:
+                    throw new Exception($"Unsupported cluster type: {cluster.Description.Type}");
+            }
+
+            var failPoints = failPointServers.Select(s => ConfigureFailPoint(s, failPoint)).Where(fp => fp != null).ToList();
```

Comment:
I would refactor the following line as:

            var session = NoCoreSession.NewHandle();
            var failPoints = failPointServers.Select(server => FailPoint.Configure(server, session, failPoint.AsBsonDocument)).ToList();

To allow getting rid of the ConfigureFailPoint method above.

Reply:
Note: the fp != null test seems to be unnecessary.